airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/report_streams.py

```diff
@@ -263,51 +268,65 @@ def _send_http_request(self, url: str, profile_id: int, json: dict = None):
             raise TooManyRequests()
         return response
 
-    @staticmethod
-    def get_report_date_ranges(start_report_date: Optional[DateTime]) -> Iterable[str]:
-        """
-        Generates dates in YYYYMMDD format for each day started from
-        start_report_date until current date (current date included)
-        :param start_report_date Starting date to generate report date list. In
-        case it is None it would return today's date.
-        :return List of days from start_report_date up until today in format
-        specified by REPORT_DATE_FORMAT variable.
-        """
-        now = pendulum.now(tz="UTC")
-        if not start_report_date:
-            start_report_date = now
-
-        # You cannot pull data for amazon ads more than 60 days
-        if (now - start_report_date).days > (60 - ReportStream.LOOK_BACK_WINDOW):
-            start_report_date = now + timedelta(days=-(60 - ReportStream.LOOK_BACK_WINDOW))
-
-        for days in range(0, (now - start_report_date).days + 1):
-            next_date = start_report_date + timedelta(days=days)
-            next_date = next_date.format(ReportStream.REPORT_DATE_FORMAT)
-            yield next_date
+    def get_report_date_ranges(self, start_report_date: Date, end_report_date: Date) -> Iterable[str]:
+        for days in range((end_report_date - start_report_date).days + 1):
+            yield start_report_date.add(days=days).format(ReportStream.REPORT_DATE_FORMAT)
+
+    def get_start_date(self, profile: Profile, stream_state: Mapping[str, Any]) -> Date:
+        today = pendulum.today(tz=profile.timezone).date()
+        if self._start_date:
+            return max(self._start_date, today.subtract(days=60))
+        start_date = stream_state.get(self.STATE_START_DATE_KEY)
+        if start_date:
+            start_date = pendulum.parse(start_date).astimezone(pytz.timezone(profile.timezone)).date()
+            return max(start_date, today.subtract(days=60))
+        return today
+
+    def get_saved_date(self, profile: Profile, stream_state: Mapping[str, Any]) -> Optional[Date]:
+        saved_date = stream_state.get(str(profile.profileId), {}).get(self.cursor_field)
+        if saved_date:
+            saved_date = pendulum.from_format(saved_date, self.REPORT_DATE_FORMAT).date()
+            today = pendulum.today(tz=profile.timezone).date()
+            if today.subtract(days=self.LOOK_BACK_WINDOW) < saved_date <= today:
+                return today.subtract(days=self.LOOK_BACK_WINDOW)
+            return saved_date
+
+    def get_start_report_date(self, profile: Profile, stream_state: Mapping[str, Any]) -> Date:
```

Comment:
can we add a comment explaining the algorithm since this is becoming quite complex

Reply:
I have re-implemented as for me general idea more simple now